source/extensions/common/wasm/context.cc

```diff
@@ -1234,32 +1234,33 @@ WasmResult Context::defineMetric(uint32_t metric_type, std::string_view name,
     return WasmResult::BadArgument;
   }
   auto type = static_cast<MetricType>(metric_type);
-  // Prefix the given name with CustomStatNamespacePrefix so that these user-defined
-  // custom metrics can be distinguished from native Envoy metrics.
-  // TODO(mathetake): Symbolize CustomStatNamespacePrefix beforehand, and pass it
-  // with the user provided StatName to Stats::Utility::*FromStatNames
-  // for less global lock time.
-  const auto prefixed_name = absl::StrCat(CustomStatNamespacePrefix, std::string(name));
   // TODO: Consider rethinking the scoping policy as it does not help in this case.
-  Stats::StatNameManagedStorage storage(prefixed_name, wasm()->scope_->symbolTable());
+  Stats::StatNameManagedStorage storage(std::string(name), wasm()->scope_->symbolTable());
   Stats::StatName stat_name = storage.statName();
+  // We prefix the given name with custom_stat_name_ so that these user-defined
+  // custom metrics can be distinguished from native Envoy metrics.
   if (type == MetricType::Counter) {
     auto id = wasm()->nextCounterMetricId();
-    auto c = &wasm()->scope_->counterFromStatName(stat_name);
+    auto c = &Stats::Utility::counterFromElements(*wasm()->scope_,
+                                                  {wasm()->custom_stat_namespace_, stat_name});
```

Comment:
s/auto/Stats::Counter*/

s/stat_name/name/ and you can delete the declaration for `storage` and the initialization of `stat_name` -- counterFromElements takes care of that.